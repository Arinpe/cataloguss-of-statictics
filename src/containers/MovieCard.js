import { Button } from '@material-ui/core';
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import setMovie from '../actions/movie';

export const MovieCard = ({
  imgUrl, handleTrailerClick, name, setMovie, id, overview, releaseDate, genres,
}) => {
  const history = useHistory();
  const handleClick = (e) => {
    e.preventDefault();
    setMovie({
      id, imgUrl, name, overview, releaseDate, genres,
    });
    history.push(`/${id}`);
  };

  return (
    <div className="movieCard">
      <div className="inner">
        <div className="front">
          <img src={imgUrl} alt="Movie Poster" />
        </div>
        <div className="back">
          <Button
            variant="outlined"
            size="large"
            color="secondary"
            onClick={() => handleTrailerClick(name)}
          >
            Trailer
          </Button>
          <Button
            variant="outlined"
            size="large"
            color="secondary"
            onClick={handleClick}
          >
            Details
          </Button>
        </div>
      </div>

    </div>
  );
};

MovieCard.defaultProps = {
  imgUrl: '',
};

MovieCard.propTypes = {
  imgUrl: PropTypes.string,
  name: PropTypes.string.isRequired,
  handleTrailerClick: PropTypes.func.isRequired,
  setMovie: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  overview: PropTypes.string.isRequired,
  releaseDate: PropTypes.string.isRequired,
  genres: PropTypes.arrayOf(PropTypes.number).isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  setMovie: (obj) => dispatch(setMovie(obj)),
});
export default connect(null, mapDispatchToProps)(MovieCard);
