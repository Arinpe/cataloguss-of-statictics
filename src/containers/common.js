import movieTrailer from 'movie-trailer';

export const generateRandomNumber = (num) => Math.floor(Math.random() * num);

export const filterMovies = (data, param) => {
  const genreId = parseInt(param, 10);
  if (genreId === 0) return data;
  return data.filter(({ genreIds }) => genreIds.includes(genreId));
};

export const getYoutubeVideoId = async (name) => {
  const url = await movieTrailer(name);
  const urlParams = new URLSearchParams(new URL(url).search);
  return urlParams.get('v');
};
