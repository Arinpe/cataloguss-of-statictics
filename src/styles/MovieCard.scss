.movieCard {
  width: 215px;
  height: 320px;
  perspective: 1000px;
  cursor: pointer;
  margin-right: 10px;

  .inner {
    position: relative;
    width: 100%;
    height: 100%;
    transform-style: preserve-3d;
    transition: transform 1s;
  }

  &:hover .inner {
    transform: rotateY(180deg);
    text-align: center;
  }

  .front,
  .back {
    position: absolute;
    width: 100%;
    height: 100%;
    backface-visibility: hidden;
  }

  .front {
    img {
      object-fit: cover;
    }
  }

  .back {
    background-color: #131313;
    transform: rotateY(180deg);
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
    align-items: center;
    border: 1px solid #fff;
    cursor: auto;
  }
}
