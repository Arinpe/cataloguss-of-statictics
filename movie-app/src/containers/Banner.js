import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Youtube from 'react-youtube';
import {
  makeStyles, Grid, Typography, Button,
} from '@material-ui/core';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { generateRandomNumber, getYoutubeVideoId } from '../helpers/common';
import { urls, imageBaseUrl, youtubeOptions } from '../helpers/constant';
import fetch from '../helpers/request';
import setMovie from '../actions/movie';

export const Banner = ({ updateMovie }) => {
  const [movie, setMovie] = useState(null);
  const [trailerId, setTrailerId] = useState('');
  const history = useHistory();
  useEffect(async () => {
    const fetchMovies = async () => {
      const {
        data: { results },
      } = await fetch.get(urls.originals);
      const index = generateRandomNumber(results.length);

      return results[index];
    };
    try {
      const data = await fetchMovies();
      setMovie(data);
    } catch {
      return null;
    }
    return false;
  }, []);

  const handleClick = () => {
    const bannerMovie = {
      name: movie.name,
      genres: movie.genre_ids,
      releaseDate: movie.first_air_date,
      overview: movie.overview,
      imgUrl: imageBaseUrl + movie.poster_path,
      id: movie.id,
    };
    updateMovie(bannerMovie);
    history.push(`/${movie.id}`);
  };

  let bgLink = '';
  if (imageBaseUrl && movie) {
    bgLink = `url('${imageBaseUrl + movie?.backdrop_path}') no-repeat center`;
  }

  const styles = makeStyles((theme) => ({
    bg: {
      background: bgLink,
      backgroundSize: 'cover',
      width: '100%',
      height: '600px',
      paddingTop: theme.spacing(6),
    },
    content: {
      height: '100%',
      paddingLeft: '30px',
      color: 'white',
    },
    btnContainer: {
      marginTop: '20px',
    },
    btn: {
      marginRight: '10px',
      background: 'white',
    },
  }));

  const handleTrailerClick = async () => {
    if (trailerId) {
      setTrailerId('');
    } else {
      try {
        const id = await getYoutubeVideoId(movie.name);
        setTrailerId(id);
      } catch {
        return null;
      }
    }
    return null;
  };

  const classes = styles();
  return trailerId ? (
    <Youtube videoId={trailerId} opts={youtubeOptions} />
  ) : (
    <div className={classes.bg}>
      <Grid container alignItems="center" className={classes.content}>
        <Grid item xs={10} md={5} lg={4}>
          <Typography variant="h4" className={classes.heading}>
            {movie?.name}
          </Typography>
          <Typography variant="body2">{movie?.overview}</Typography>
          <div className={classes.btnContainer}>
            <Button
              className={classes.btn}
              variant="outlined"
              size="large"
              color="secondary"
              onClick={handleTrailerClick}
            >
              Play
            </Button>
            <Button
              className={classes.btn}
              variant="outlined"
              size="large"
              color="secondary"
              onClick={handleClick}
            >
              Details
            </Button>
          </div>
        </Grid>
      </Grid>
      {trailerId && <Youtube videoId={trailerId} opts={youtubeOptions} />}
    </div>
  );
};

Banner.propTypes = {
  updateMovie: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  updateMovie: (obj) => dispatch(setMovie(obj)),
});

export default connect(null, mapDispatchToProps)(Banner);
