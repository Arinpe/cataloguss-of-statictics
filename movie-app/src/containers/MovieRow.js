import { Grid, makeStyles } from '@material-ui/core';
import React, { useEffect, useState, useRef } from 'react';
import { connect } from 'react-redux';
import Youtube from 'react-youtube';
import PropTypes from 'prop-types';
import Card from './MovieCard';
import Filter from '../components/Filter';
import fetch from '../helpers/request';
import { imageBaseUrl, youtubeOptions } from '../helpers/constant';
import { filterMovies, getYoutubeVideoId } from '../helpers/common';
import filterAction from '../actions/filter';

const useStyles = makeStyles((theme) => ({
  itemRoot: {
    padding: '0 !important',
  },
  container: {
    flexWrap: 'nowrap',
    'overflow-x': 'scroll',
    padding: theme.spacing(2, 0),
  },
  heading: {
    padding: '15px 0',
  },
}));
function MovieRow({
  title, moviesUrl, filterParam, moviesFilter, id, allGenres,
}) {
  const classes = useStyles();
  const [movies, setMovies] = useState(null);
  const parentSectionId = useRef('');
  const [genres, setGenres] = useState([]);
  const [trailerId, setTrailerId] = useState('');
  useEffect(async () => {
    const fetchMovies = async () => {
      const { data: { results } } = await fetch.get(moviesUrl);

      return results.map((movie) => ({
        id: movie.id,
        imageUrl: imageBaseUrl + movie.poster_path,
        genreIds: movie.genre_ids,
        name: movie.title || movie.name,
        overview: movie.overview,
        releaseDate: movie.first_air_date || movie.release_date,
      }));
    };

    try {
      const data = await fetchMovies();
      setMovies(data);
      const flattedGenresArr = data.map(({ genreIds }) => (genreIds)).flat();
      const uniqGenres = new Set(flattedGenresArr);
      setGenres([...uniqGenres]);
      return true;
    } catch {
      return null;
    }
  }, []);

  const setMoviesToDisplay = () => {
    if (parentSectionId.current === id) {
      return filterMovies(movies, filterParam);
    }
    return movies;
  };

  const handleFilter = (e) => {
    const section = e.target.closest('section');
    if (id === section.id) {
      moviesFilter(e.target.value);
      parentSectionId.current = section.id;
    }
  };

  const handleTrailerClick = async (name) => {
    if (trailerId) {
      setTrailerId('');
    } else {
      try {
        const id = await getYoutubeVideoId(name);
        setTrailerId(id);
      } catch {
        return null;
      }
    }
    return true;
  };
  const allMovies = filterParam.toString() === '0' ? movies : setMoviesToDisplay();
  return (
    <div className="container">
      <section id={id}>
        <Grid className={classes.heading} container justifyContent="space-between" alignItems="center">
          <Grid item>
            <h5 className="section__heading">{title}</h5>
          </Grid>
          <Grid item>
            <Filter
              allGenres={allGenres}
              handleFilter={handleFilter}
              genres={genres}
            />
          </Grid>
        </Grid>
        <Grid className={`${classes.container} scrollbar`} container spacing={3}>
          {allMovies?.map(({
            id, imageUrl, name, overview, releaseDate, genreIds,
          }) => (
            <Grid className={classes.itemRoot} item key={id}>
              <Card
                id={id}
                name={name}
                handleTrailerClick={handleTrailerClick}
                imgUrl={imageUrl}
                desc={overview}
                releaseDate={releaseDate}
                genres={genreIds}
                overview={overview}
              />
            </Grid>
          ))}
        </Grid>
      </section>
      {trailerId && <Youtube videoId={trailerId} opts={youtubeOptions} />}
    </div>
  );
}

MovieRow.propTypes = {
  title: PropTypes.string.isRequired,
  moviesUrl: PropTypes.string.isRequired,
  filterParam: PropTypes.number.isRequired,
  moviesFilter: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  allGenres: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  })).isRequired,
};

const mapStateToProps = (state) => ({
  filterParam: parseInt(state.filter, 10),
});

const mapDispatchToProps = (dispatch) => ({
  moviesFilter: (param) => dispatch(filterAction(param)),
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieRow);
