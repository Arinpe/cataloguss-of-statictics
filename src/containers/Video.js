import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom';
import Youtube from 'react-youtube';
import { format } from 'date-fns';
import { getYoutubeVideoId } from '../helpers/common';

export const Video = ({ movie, allGenres }) => {
  const history = useHistory();
  if (!movie || !Object.keys(movie).length) {
    history.push('/');
  }
  const [trailerId, setTrailerId] = useState('null');
  const {
    name, imgUrl, overview, releaseDate, genres,
  } = movie;

  useEffect(async () => {
    try {
      const videoId = await getYoutubeVideoId(name);
      setTrailerId(videoId);
    } catch {
      return null;
    }
    return false;
  });

  const genresArr = genres?.map((id) => (
    allGenres?.find((gen) => gen.id === id)
  )).filter((item) => item);

  const opts = {
    height: '500px',
    width: '100%',
    autoplay: 1,
  };

  return (
    <div className="video">
      <div className="video__wrapper">
        <div className="d-flex video__header">
          <h2>{name}</h2>
          <span>
            (
            {releaseDate?.split('-')[0]}
            )
          </span>
        </div>
        <div className="video__content">
          <div className="d-flex video__media">
            <div className="video__poster">
              <img src={imgUrl} alt="Movie Poster" />
            </div>
            <div className="video__trailer">
              {trailerId && <Youtube videoId={trailerId} opts={opts} />}
            </div>
          </div>
        </div>
        <div className="video__description">
          <p>{overview}</p>
        </div>
        <div className="video__footer">
          <div>
            <p>Release date</p>
            <p>{releaseDate && format(new Date(releaseDate), 'd LLLL yyyy')}</p>
          </div>
          <ul className="video__genres">
            {genresArr?.map(({ name, id }) => <li key={id}>{name}</li>)}
          </ul>
        </div>
      </div>
    </div>
  );
};

Video.propTypes = {
  movie: PropTypes.shape({
    imgUrl: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    overview: PropTypes.string.isRequired,
    releaseDate: PropTypes.string.isRequired,
    genres: PropTypes.arrayOf(PropTypes.number).isRequired,
  }).isRequired,
  allGenres: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  })).isRequired,
};

const mapStateToProps = ({ movie, genres }) => ({
  movie,
  allGenres: genres,
});

export default connect(mapStateToProps)(Video);
