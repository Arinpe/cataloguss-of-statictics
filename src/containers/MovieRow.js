import { Grid } from '@material-ui/core';
import React, { useEffect, useState, useRef } from 'react';
import { connect } from 'react-redux';
import Youtube from 'react-youtube';
import PropTypes from 'prop-types';
import Card from './MovieCard';
import Filter from '../components/filter';
import fetch from '../helpers/request';
import { imageBaseUrl, youtubeOptions } from '../helpers/constant';
import { filterMovies, getYoutubeVideoId } from '../helpers/common';
import filterAction from '../actions/filter';
import { movieRowStyles } from '../helpers/styles';

function MovieRow({
  title, moviesUrl, filterParam, moviesFilter, id,
}) {
  const classes = movieRowStyles();
  const [movies, setMovies] = useState(null);
  const parentSectionId = useRef('');
  const [trailerId, setTrailerId] = useState('');
  useEffect(async () => {
    const fetchMovies = async () => {
      const { data: { results } } = await fetch.get(moviesUrl);

      return results.map((movie) => ({
        id: movie.id,
        imageUrl: imageBaseUrl + movie.poster_path,
        genreIds: movie.genre_ids,
        name: movie.title || movie.name,
        overview: movie.overview,
        releaseDate: movie.first_air_date || movie.release_date,
      })).sort((a, b) => a - b);
    };

    try {
      const data = await fetchMovies();
      setMovies(data);
      return true;
    } catch {
      return null;
    }
  }, []);

  const setMoviesToDisplay = () => {
    if (parentSectionId.current === id) {
      return filterMovies(movies, filterParam);
    }
    return movies;
  };

  const handleFilter = (e) => {
    const section = e.target.closest('section');
    if (id === section.id) {
      moviesFilter(e.target.value);
      parentSectionId.current = section.id;
    }
  };

  const handleTrailerClick = async (name) => {
    if (trailerId) {
      setTrailerId('');
    } else {
      try {
        const id = await getYoutubeVideoId(name);
        setTrailerId(id);
      } catch {
        return null;
      }
    }
    return true;
  };
  const allMovies = filterParam.toString() === '0' ? movies : setMoviesToDisplay();
  return (
    <div className="container">
      <section id={id}>
        <Grid className={classes.heading} container justifyContent="space-between" alignItems="center">
          <Grid item>
            <h5 className="section__heading">{title}</h5>
          </Grid>
          <Grid item>
            <Filter
              handleFilter={handleFilter}
            />
          </Grid>
        </Grid>
        <Grid className={`${classes.container} scrollbar`} container spacing={3}>
          {allMovies?.map(({
            id, imageUrl, name, overview, releaseDate, genreIds,
          }) => (
            <Grid className={classes.itemRoot} item key={id}>
              <Card
                id={id}
                name={name}
                handleTrailerClick={handleTrailerClick}
                imgUrl={imageUrl}
                desc={overview}
                releaseDate={releaseDate}
                genres={genreIds}
                overview={overview}
              />
            </Grid>
          ))}
        </Grid>
      </section>
      {trailerId && <Youtube videoId={trailerId} opts={youtubeOptions} />}
    </div>
  );
}

MovieRow.propTypes = {
  title: PropTypes.string.isRequired,
  moviesUrl: PropTypes.string.isRequired,
  filterParam: PropTypes.number.isRequired,
  moviesFilter: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
  filterParam: parseInt(state.filter, 10),
});

const mapDispatchToProps = (dispatch) => ({
  moviesFilter: (param) => dispatch(filterAction(param)),
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieRow);
