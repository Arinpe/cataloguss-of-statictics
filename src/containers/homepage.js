import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import MovieBanner from './Banner';
import MovieRow from './MovieRow';
import { urls } from '../helpers/constant';
import fetchData from '../helpers/request';
import setGenres from '../actions/genres';

const HomePage = ({ genres, setGenres }) => {
  useEffect(async () => {
    const fetchGenres = async () => {
      const { data: genres } = await fetchData.get(urls.genresUrl);
      return genres.genres;
    };
    const res = await fetchGenres();
    setGenres(res);
  }, []);
  return (
    <div>
      <MovieBanner />
      <MovieRow allGenres={genres} id="popularMovies" title="Popular Movies" moviesUrl={urls.moviePopular} />
      <MovieRow allGenres={genres} id="popularTv" title="Popular Tv Shows" moviesUrl={urls.popularTv} />
      <MovieRow allGenres={genres} id="topRatedMovies" title="Top Rated Movies" moviesUrl={urls.topRatedMovie} />
      <MovieRow allGenres={genres} id="topRatedTv" title="Top Rated Tv Shows" moviesUrl={urls.topRatedTv} />
      <MovieRow allGenres={genres} id="upcoming" title="Upcoming" moviesUrl={urls.upcoming} />
    </div>
  );
};

HomePage.propTypes = {
  genres: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  })).isRequired,
  setGenres: PropTypes.func.isRequired,
};

const mapStateToProps = ({ genres }) => ({
  genres,
});

const mapDispatchToProps = (dispatch) => ({
  setGenres: (genres) => dispatch(setGenres(genres)),
});

export default connect(mapStateToProps, mapDispatchToProps)(HomePage);
