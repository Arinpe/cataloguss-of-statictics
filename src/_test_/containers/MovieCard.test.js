import { Button } from '@material-ui/core';
import { MovieCard } from '../../containers/MovieCard';
import shallowWrapper from '../testSetup';

describe('Movie Card', () => {
  const movieInfo = {
    imgUrl: 'imageUrl',
    name: 'Test',
    handleTrailerClick: jest.fn(),
    setMovie: jest.fn(),
    id: 3,
    overview: 'Movie Description',
    releaseDate: '2020-12-10',
    genres: [1, 2, 3, 4],
  };

  const card = shallowWrapper(MovieCard, movieInfo);

  test('should render', () => {
    expect(card).toBeTruthy();
    expect(card).not.toBeFalsy();
    expect(card).toMatchSnapshot();
  });

  test('should have a image url', () => {
    const image = card.find('img');
    expect(image.prop('src')).toEqual(movieInfo.imgUrl);
    expect(image.prop('src')).not.toEqual('');
  });

  test('should render Trailer button', () => {
    const buttons = card.find(Button);

    expect(buttons).toHaveLength(2);
    expect(buttons).not.toHaveLength(0);
    expect(buttons.first().text()).toContain('Trailer');
    expect(buttons.first().text()).not.toBeNull();
  });

  test('should render details button', () => {
    const buttons = card.find(Button);

    expect(buttons).toHaveLength(2);
    expect(buttons.last().text()).toContain('Details');
    expect(buttons.last().text()).not.toBeNull();
  });
});
