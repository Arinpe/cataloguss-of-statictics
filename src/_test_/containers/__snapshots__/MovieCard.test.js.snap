// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Card should render 1`] = `ShallowWrapper {}`;
